From 0da5a5b413717d81aca911bfbf43f88663c95368 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Sat, 9 Apr 2022 15:48:53 +0300
Subject: [PATCH 2/4] gmscompat: restart GMS processes when permission gets
 granted

See frameworks_base commit for tags that can be used to see the full commit history.
---
 .../permission/utils/KotlinUtils.kt               | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/PermissionController/src/com/android/permissioncontroller/permission/utils/KotlinUtils.kt b/PermissionController/src/com/android/permissioncontroller/permission/utils/KotlinUtils.kt
index 2aad0154bc..9d1e90aadc 100644
--- a/PermissionController/src/com/android/permissioncontroller/permission/utils/KotlinUtils.kt
+++ b/PermissionController/src/com/android/permissioncontroller/permission/utils/KotlinUtils.kt
@@ -35,6 +35,7 @@ import android.app.AppOpsManager.MODE_IGNORED
 import android.app.AppOpsManager.OPSTR_AUTO_REVOKE_PERMISSIONS_IF_UNUSED
 import android.app.AppOpsManager.permissionToOp
 import android.app.Application
+import android.app.compat.gms.GmsCompat
 import android.content.Context
 import android.content.Intent
 import android.content.Intent.ACTION_MAIN
@@ -1066,6 +1067,20 @@ object KotlinUtils {
             )
         }
 
+        if (GmsCompat.isEnabledFor(pkgInfo.packageName, user.identifier)) {
+            // in many cases, GMS needs a restart to properly handle permission grants
+            if (
+                // Google Search app handles permission grants properly.
+                // (GmsInfo.PACKAGE_GSA is inaccessible here)
+                pkgInfo.packageName != "com.google.android.googlequicksearchbox"
+                // Play Store asks for POST_NOTIFICATIONS on first launch
+                && perm.name != POST_NOTIFICATIONS
+                && perm.name != Manifest.permission.OTHER_SENSORS
+            ) {
+                shouldKill = true
+            }
+        }
+
         val newState = PermState(newFlags, isGranted)
         return LightPermission(perm.pkgInfo, perm.permInfo, newState, perm.foregroundPerms) to
             shouldKill
-- 
2.34.1

